aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,338 @@
+///   * The object model allows for events to be directly coupled with data.
+/// * Objects definition / structure can prevented after creation
+///   * At creation, the creator has the ability to configure the account as desired
+///   * Afterward, the creator can delegate signer ability to continue to change definition
+///     but it can also be discarded or access prevented to ensure that an object maintains
+///     consistent shape.
+/// * An object should be highly configurable, the creator the following access privileges:
+///   * Ownership
+///   * Signer
+///   * Deletion
+///   * Mutation
+///   * Reading
+/// * Secure and guaranteed to not collide with existing accounts.
+///   * Each address is generated in a collision-resistant fashion guaranteeing legitimate
+///     accounts cannot be created within the same address.
+///   * Two objects cannot be co-located due to an address being able to only store a single Object
+///
+/// TODO:
+/// * Currently an object will be spread across distinct storage slots, this is not efficient.
+///   We are currently exploring a concept of storage groups that will allow structs marked with
+///   `[storage_group(GROUP)]` to be co-located in a storage slot.
+/// * There is no means to borrow an object or a reference to an object. We are exploring how to
+///   make it so that a reference to a global object can be returned from a function.
+/// * There's no guarantee of correctness or consistency in object implementation. For example,
+///   the API for generating a TypedOwnerAbility can differ from object to object. An object
+///   could potentially be left in a partially deleted state. We are exploring interfaces that
+///   would allow for calling functions on module, type pairs that implement certain functions.
+/// * The ownership model allows for richer defintion of access control, such as specifying when
+///   an object can be mutated, read, and moved. This can be extended to entry functions to
+///   seamlessly allow a user to know prior to sending a transaction the potential implications
+///   for their owned objects.
+/// * The current demo is rather limited and needs to be extended to demonstrate more end-to-end.
+module aptos_framework::object {
+    use std::bcs;
+    use std::error;
+    use std::hash;
+    use std::signer;
+    use std::vector;
+
+    use aptos_std::table;
+
+    use aptos_framework::account;
+    use aptos_framework::event;
+    use aptos_framework::from_bcs;
+    use aptos_framework::guid;
+
+    /// An object already exists at this address
+    const EOBJECT_EXISTS: u64 = 0;
+    /// An object of that type already exists at this address
+    const EOBJECT_TYPE_EXISTS: u64 = 1;
+
+    /// Scheme identifier used to generate an object's address. This serves as domain separation to
+    /// prevent existing authentication key and resource account derivation to produce an object
+    /// address.
+    const OBJECT_ADDRESS_SCHEME: u8 = 0xFE;
```

Comment:
Since this is going to be hashed anyway, why not just use bytes, e.g. b"object"?

Reply:
This is the paradigm we use in address generation. @alinush